tmk_core/common/action_layer.c

```diff
@@ -301,9 +301,9 @@ uint8_t layer_switch_get_layer(keypos_t key) {
   action_t action;
   action.code = ACTION_TRANSPARENT;
 
-  uint32_t layers = layer_state | default_layer_state;
+  layer_state_t layers = layer_state | default_layer_state;
   /* check top layer first */
-  for (int8_t i = 31; i >= 0; i--) {
+  for (int8_t i = sizeof(layer_state_t)-1; i >= 0; i--) {
```

Comment:
This is probably the issue. `sizeof(layer_state_t)-1` returns 3, not 31 as this is the size in bytes of the type. This should be  `sizeof(layer_state_t) * 8 - 1`

Reply:
Can confirm changing this as described fixes layer switching.